src/build123d/topology.py

```diff
@@ -8592,7 +8379,58 @@ def chamfer_2d(
         Returns:
             Wire: chamfered wire
         """
-        return Face(self).chamfer_2d(distance, distance2, vertices, edge).outer_wire()
+        reference_edge = edge
+        del edge
+
+        # Create a face to chamfer
+        unchamfered_face = _make_topods_face_from_wires(self.wrapped)
+        chamfer_builder = BRepFilletAPI_MakeFillet2d(unchamfered_face)
+
+        vertex_edge_map = TopTools_IndexedDataMapOfShapeListOfShape()
+        TopExp.MapShapesAndAncestors_s(
+            unchamfered_face, ta.TopAbs_VERTEX, ta.TopAbs_EDGE, vertex_edge_map
+        )
+
+        for v in vertices:
+            edges = vertex_edge_map.FindFromKey(v.wrapped)
+
+            # Index or iterator access to OCP.TopTools.TopTools_ListOfShape is slow on M1 macs
+            # Using First() and Last() to omit
+            edges = [edges.First(), edges.Last()]
+
+            # Need to wrap in b3d objects for comparison to work
+            # ref.wrapped != edge.wrapped but ref == edge
+            edges = [Edge(e) for e in edges]
+
+            edge1, edge2 = Wire.order_chamfer_edges(reference_edge, edges)
+
+            chamfer_builder.AddChamfer(
+                TopoDS.Edge_s(edge1.wrapped),
+                TopoDS.Edge_s(edge2.wrapped),
+                distance,
+                distance2,
+            )
+
+        chamfer_builder.Build()
+        chamfered_face = chamfer_builder.Shape()
+        # Fix the shape
+        shape_fix = ShapeFix_Shape(chamfered_face)
+        shape_fix.Perform()
+        chamfered_face = downcast(shape_fix.Shape())
+        # Return the outer wire
+        return Wire(BRepTools.OuterWire_s(chamfered_face))
+
+    @staticmethod
+    def order_chamfer_edges(reference_edge, edges) -> tuple[Edge, Edge]:
+        """Order the edges of a chamfer relative to a reference Edge"""
+        if reference_edge:
+            if reference_edge not in edges:
+                raise ValueError("One or more vertices are not part of edge")
+            edge1 = reference_edge
+            edge2 = [x for x in edges if x != reference_edge][0]
+        else:
+            edge1, edge2 = edges
+        return edge1, edge2
```

Comment:
assuming `edges` is always a pair not an arbitrary list
```suggestion
    def order_chamfer_edges(reference_edge: Optional[Edge], edges: tuple[Edge, Edge]) -> tuple[Edge, Edge]:
        """Order the edges of a chamfer relative to a reference Edge"""
        if reference_edge:
            edge1, edge2 = edges
            if edge1 == reference_edge:
                return edge1, edge2
            elif edge2 == reference_edge:
                return edge2, edge1
            else:
                raise ValueError("reference edge not in edges")
        else:
            return edges
```

Reply:
Nice - used your version.